docs/tutorials/jaxpr.md

```diff
@@ -0,0 +1,290 @@
+
+## `jax.core.ClosedJaxpr`
+
+A jaxpr instance represents a function with one or more typed parameters (input variables) and one or more typed results. The results depend only on the input variables; there are no free variables captured from enclosing scopes. The inputs and outputs have types, which in JAX are represented as abstract values.
+
+There are two related representations in the code for jaxprs, {class}`jax.core.Jaxpr` and {class}`jax.core.ClosedJaxpr`. A {class}`jax.core.ClosedJaxpr` represents a partially-applied {class}`jax.core.Jaxpr`, and is what you obtain when you use {func}`jax.make_jaxpr` to inspect jaxprs. It has the following fields:
+
+- `jaxpr`: is a {class}`jax.core.Jaxpr` representing the actual computation content of the function (described below).
+- `consts` is a list of constants.
+
+The most interesting part of the `ClosedJaxpr` is the actual execution content, represented as a {class}`jax.core.Jaxpr` as printed using the following grammar:
+
+```
+jaxpr ::= { lambda Var* ; Var+.
+            let Eqn*
+            in  [Expr+] }
+```
+
+where:
+
+- The parameters of the jaxpr are shown as two lists of variables separated by `;`:
+    - The first set of variables are the ones that have been introduced to stand for constants that have been hoisted out. These are called the `constvars`, and in a {class}`jax.core.ClosedJaxpr` the `consts` field holds corresponding values.
+    - The second list of variables, called `invars`, correspond to the inputs of the traced Python function.
+- `Eqn*` is a list of equations, defining intermediate variables referring to intermediate expressions. Each equation defines one or more variables as the result of applying a primitive on some atomic expressions. Each equation uses only input variables and intermediate variables defined by previous equations.
+- `Expr+`: is a list of output atomic expressions (literals or variables) for the jaxpr.
+
+Equations are printed as follows:
+
+```
+Eqn  ::= let Var+ = Primitive [ Param* ] Expr+
+```
+
+where:
+
+- `Var+` are one or more intermediate variables to be defined as the output of a primitive invocation (some primitives can return multiple values).
+- `Expr+` are one or more atomic expressions, each either a variable or a literal constant. A special variable `unitvar` or literal `unit`, printed as `*`, represents a value that is not needed in the rest of the computation and has been elided. That is, units are just placeholders.
+- `Param*` are zero or more named parameters to the primitive, printed in square brackets. Each parameter is shown as `Name = Value`.
+
+Most jaxpr primitives are first-order (they take just one or more Expr as arguments):
+
+```
+Primitive := add | sub | sin | mul | ...
+```
+
+The jaxpr primitives are documented in the {mod}`jax.lax` module.
```

Comment:
This is no longer totally true; I'd say "The most common jaxpr primitives are documented..."

Reply:
On it.